hal/device.h

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (C) 2019-2020 BiiLabs Co., Ltd. and Contributors
+ * All Rights Reserved.
+ * This is free software; you can redistribute it and/or modify it under the
+ * terms of the MIT license. A copy of the license can be found in the file
+ * "LICENSE" at the root of this distribution.
+ */
+
+#ifndef HAL_DEVICE_H
+#define HAL_DEVICE_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <stdint.h>
+#include "defined_error.h"
+
+/**
+ * Device abstract layer
+ * Implement your device inside devices
+ */
+typedef struct device_s device_t;
+struct device_s {
+  void (*destructor)(device_t *);
+  retcode_t (*get_key)(uint8_t *key);                    /**< get device private key */
+  retcode_t (*get_hardware_identifier)(char *device_id); /**< get device id */
```

Comment:
Change to `get_device_id`.

Reply:
We have to encapsulate hardware-specific operations as possible as we can. @HowJMay, I need your input.